jax/_src/lax/parallel.py

```diff
@@ -355,6 +361,81 @@ def bind(leaf):
+
+  If ``x`` is a pytree then the result is equivalent to mapping this function to
+  each leaf in the tree.
+
+  This function is an analog of the Send HLO.
+
+  Args:
+    x: array(s) with a mapped axis named ``axis_name``.
+    axis_name: hashable Python object used to name a pmapped axis (see the
+      :func:`jax.pmap` documentation for more details).
+    perm: list of pairs of ints, representing ``(source_index,
+      destination_index)`` pairs that encode how the mapped axis named
+      ``axis_name`` should be shuffled. The integer values are treated as
+      indices into the mapped axis ``axis_name``. Any two pairs should not have
+      the same source index or the same destination index. For each index of the
+      axis ``axis_name`` that does not correspond to a destination index in
+      ``perm``, the corresponding values in the result are filled with zeros of
+      the appropriate type. The semantics here are platform-specific, and for
+      GPU they correspond to NCCL send.
+
+  Returns:
+    A compiler token that can be used by precv and lax.optimzation_barrier to
+    enforce ordering of collective ops.
+  """
+  if not isinstance(axis_name, (list, tuple)):
+    axis_name = (axis_name,)
+
+  def bind(leaf):
+    leaf = insert_collective_pvary(axis_name, leaf)
+    return psend_p.bind(leaf, axis_name=axis_name, perm=tuple(map(tuple, perm)))
+
+  return tree_util.tree_map(bind, x)
+
+
+def precv(token, expected_dtype_and_shape, axis_name, perm):
+  """Perform a collective recv according to the permutation ``perm``.
+
+  This function is an analog of the Recv HLO.
+
+  Args:
+    token: a compiler token, either generated by a matching psend or
+      lax.create_token(). This is used to enforce control dependencies between
+      collectives.
+    expected_dtype_and_shape: ShapeDtypeStruct(s) containing the dtype and shape
+      of the result.
+    axis_name: hashable Python object used to name a pmapped axis (see the
+      :func:`jax.pmap` documentation for more details).
+    perm: list of pairs of ints, representing ``(source_index,
+      destination_index)`` pairs that encode how the mapped axis named
+      ``axis_name`` should be shuffled. The integer values are treated as
+      indices into the mapped axis ``axis_name``. Any two pairs should not have
+      the same source index or the same destination index. For each index of the
+      axis ``axis_name`` that does not correspond to a destination index in
+      ``perm``, the corresponding values in the result are filled with zeros of
+      the appropriate type. The semantics here are platform-specific, and for
+      GPU they correspond to NCCL recv.
+
+  Returns:
+    Array(s) with the same shape as ``expected_dtype_and_shape``.
+  """
+  if not isinstance(axis_name, (list, tuple)):
+    axis_name = (axis_name,)
+  return precv_p.bind(
+      token,
+      expected_dtype_and_shape=core.ShapedArray(
```

Comment:
why convert this to a ShapedArray?

Reply:
I think it makes sense: we accept ducks but convert to something more precise in the jaxpr, in particular exactly the thing we need for the abstract eval rule.

Reply:
+1 to what Matt said